Make types appear explicitly in input files

nilsvu opened the pull request:
## Proposed changes

This PR is a follow-up to #1483, making use of the new option groups feature to add more structure to our input files. Mainly, it proposes the following change to how options for classes chosen at compile-time appear in the input file:

**Old:**
```yaml
AnalyticSolution:
  BhMass: 1.0
  ...
```

**New:**
```yaml
AnalyticSolution:
  FishboneMoncriefDisk:
    BhMass: 1.0
    ...
```

Note that this is backwards-compatible: If we choose to allow specifying the analytic solution in the input file in the future, as we do for the time stepper, for instance, then no changes to the input file are needed. On the other hand, if the user tries to change the `FishboneMoncriefDisk` to something else without the executable supporting it, a standard parse error is shown.

The reason for this change I also commented on in https://github.com/sxs-collaboration/spectre/pull/1483#issuecomment-482456502.

- [x] Depends on #1483, so the first 7 commits are not part of this PR

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

nilsvu committed:
Remove "Params" suffix from option tags

nilsvu committed:
Make options name function public

nilsvu committed:
Make analytic solution type appear in input file

nilsvu committed:
Make limiter type appear in input file

nilsvu committed:
Make numerical flux type appear in input file

nilsvu committed:
Make the evolution system appear in the input file

fmahebert commented:
I like this improvement to the human-readability of the input files :+1:

wthrowe commented on src/Options/Options.hpp:
```diff
@@ -145,3 +147,25 @@ struct create_from_yaml {
+// The name in the YAML file for a struct.
+template <typename T>
+std::string option_name() noexcept {
+  return Options_detail::name<T>();
```
Use `name_helper` directly and get rid of the `name` detail function.  (Everything can just call `option_name`.)

wthrowe reviewed (changes requested):
I like these changes as well.  I just have one minor code request.  (Also Travis is failing, but I think that's because of the PR this is on top of, and I see you're working on that.)

nilsvu commented:
Fixed @wthrowe. Also rebased on developed since #1483 was merged.

@nilsdeppe do you want to review this, since you brought up some points in #1482?

nilsdeppe commented:
Yes, I'll look later today or tomorrow :)

nilsdeppe commented on src/PointwiseFunctions/AnalyticData/Tags.hpp:
```diff
@@ -6,6 +6,12 @@
 #include "Options/Options.hpp"
 
 namespace OptionTags {
+struct AnalyticDataGroup {
```
I think you documented some of the other groups already, so let's document these ones too

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Tags.hpp:
```diff
@@ -6,14 +6,21 @@
 #include "Options/Options.hpp"
 
 namespace OptionTags {
+struct SlopeLimiterGroup {
```
doxy

nilsdeppe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/Tags.hpp:
```diff
@@ -50,14 +50,23 @@ struct MortarSize : db::SimpleTag {
 }  // namespace Tags
 
 namespace OptionTags {
+struct NumericalFluxGroup {
```
doxy?

nilsdeppe commented on src/Evolution/Systems/GrMhd/ValenciaDivClean/Tags.hpp:
```diff
@@ -9,16 +9,24 @@
 class DataVector;
 
 namespace OptionTags {
+struct ValenciaDivCleanGroup {
```
doxy?

nilsdeppe commented on src/Evolution/Tags.hpp:
```diff
@@ -19,4 +19,9 @@ struct EvolutionGroup {
   static constexpr OptionString help{"Options for the time evolution"};
 };
 
+struct EvolutionSystemGroup {
```
doxy?

nilsdeppe reviewed (commented):
LGTM, just some documentation comments :)

nilsvu commented:
@nilsdeppe @wthrowe could you check the fixup please?

wthrowe commented:
Looks good.

nilsdeppe commented:
fine by me

nilsvu commented:
@wthrowe could you look at this and merge if you're happy with it?
